File: lib/spiral/date_time.ts
import { Union } from "../../deps/Fable/src/fable-library-ts/Types.js";
import { union_type, string_type, TypeInfo } from "../../deps/Fable/src/fable-library-ts/Reflection.js";
import { parse as parse_1, toString } from "../../deps/Fable/src/fable-library-ts/Date.js";
import { int64ToString, defaultOf } from "../../deps/Fable/src/fable-library-ts/Util.js";
import { toString as toString_1, hours } from "../../deps/Fable/src/fable-library-ts/TimeSpan.js";
import { int32, uint8 } from "../../deps/Fable/src/fable-library-ts/Int32.js";
import { newGuid, parse } from "../../deps/Fable/src/fable-library-ts/Guid.js";
import { replace, concat } from "../../deps/Fable/src/fable-library-ts/String.js";
import { toInt64, int64 } from "../../deps/Fable/src/fable-library-ts/BigInt.js";
import { parse as parse_2 } from "../../deps/Fable/src/fable-library-ts/Long.js";

export type US0_$union = 
    | US0<0>
    | US0<1>

export type US0_$cases = {
    0: ["US0_0", [any]],
    1: ["US0_1", []]
}

export function US0_US0_0(f0_0: any) {
    return new US0<0>(0, [f0_0]);
}

export function US0_US0_1() {
    return new US0<1>(1, []);
}

export class US0<Tag extends keyof US0_$cases> extends Union<Tag, US0_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: US0_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["US0_0", "US0_1"];
    }
}

export function US0_$reflection(): TypeInfo {
    return union_type("Date_time.US0", [], US0, () => [[["f0_0", string_type]], []]);
}

export type US1_$union = 
    | US1<0>
    | US1<1>

export type US1_$cases = {
    0: ["US1_0", [string]],
    1: ["US1_1", []]
}

export function US1_US1_0(f0_0: string) {
    return new US1<0>(0, [f0_0]);
}

export function US1_US1_1() {
    return new US1<1>(1, []);
}

export class US1<Tag extends keyof US1_$cases> extends Union<Tag, US1_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: US1_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["US1_0", "US1_1"];
    }
}

export function US1_$reflection(): TypeInfo {
    return union_type("Date_time.US1", [], US1, () => [[["f0_0", string_type]], []]);
}

export type US2_$union = 
    | US2<0>
    | US2<1>
    | US2<2>

export type US2_$cases = {
    0: ["US2_0", []],
    1: ["US2_1", []],
    2: ["US2_2", []]
}

export function US2_US2_0() {
    return new US2<0>(0, []);
}

export function US2_US2_1() {
    return new US2<1>(1, []);
}

export function US2_US2_2() {
    return new US2<2>(2, []);
}

export class US2<Tag extends keyof US2_$cases> extends Union<Tag, US2_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: US2_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["US2_0", "US2_1", "US2_2"];
    }
}

export function US2_$reflection(): TypeInfo {
    return union_type("Date_time.US2", [], US2, () => [[], [], []]);
}

export type US3_$union = 
    | US3<0>
    | US3<1>
    | US3<2>
    | US3<3>
    | US3<4>

export type US3_$cases = {
    0: ["US3_0", [US2_$union]],
    1: ["US3_1", [US2_$union]],
    2: ["US3_2", [US2_$union]],
    3: ["US3_3", [US2_$union]],
    4: ["US3_4", [US2_$union]]
}

export function US3_US3_0(f0_0: US2_$union) {
    return new US3<0>(0, [f0_0]);
}

export function US3_US3_1(f1_0: US2_$union) {
    return new US3<1>(1, [f1_0]);
}

export function US3_US3_2(f2_0: US2_$union) {
    return new US3<2>(2, [f2_0]);
}

export function US3_US3_3(f3_0: US2_$union) {
    return new US3<3>(3, [f3_0]);
}

export function US3_US3_4(f4_0: US2_$union) {
    return new US3<4>(4, [f4_0]);
}

export class US3<Tag extends keyof US3_$cases> extends Union<Tag, US3_$cases[Tag][0]> {
    constructor(readonly tag: Tag, readonly fields: US3_$cases[Tag][1]) {
        super();
    }
    cases() {
        return ["US3_0", "US3_1", "US3_2", "US3_3", "US3_4"];
    }
}

export function US3_$reflection(): TypeInfo {
    return union_type("Date_time.US3", [], US3, () => [[["f0_0", US2_$reflection()]], [["f1_0", US2_$reflection()]], [["f2_0", US2_$reflection()]], [["f3_0", US2_$reflection()]], [["f4_0", US2_$reflection()]]]);
}

export function closure2(unitVar: void, v0_1: any): US0_$union {
    return US0_US0_0(v0_1);
}

export function method1(): ((arg0: any) => US0_$union) {
    return (v: any): US0_$union => closure2(undefined, v);
}

export function method2(): string {
    return "hh:mm";
}

export function method3(): string {
    return "yyyyMMdd-HHmm-ssff-ffff-f";
}

export function method4(): string {
    return "hhmm";
}

export function method0(v0_1: string, v1_1: Date): string {
    const v792: string = method3();
    const v797: string = toString(v1_1, (v792 === "") ? "M-d-y hh:mm:ss tt" : v792);
    const v820: any = defaultOf();
    let v978: number;
    throw new Error(`date_time.get_utc_offset / target: ${US3_US3_3(US2_US2_0())}`);
    const v991: uint8 = (hours(v978) > 0) ? 1 : 0;
    const v992: string = method4();
    const v1016 = `${v991}${toString_1(v978, "c", {})}`;
    const v1018: string = v0_1;
    const v1028: int32 = (v797.length + v1016.length) | 0;
    const v1042: int32 = (v1018.length - 1) | 0;
    return parse(concat(v797, v1016, ...v1018.slice(v1028, v1042 + 1)));
}

export function closure1(v0_1: string, v1_1: Date): string {
    return method0(v0_1, v1_1);
}

export function closure0(unitVar: void, v0_1: string): ((arg0: Date) => string) {
    return (v: Date): string => closure1(v0_1, v);
}

export function closure4(unitVar: void, v0_1: string): string {
    return replace(v0_1, "-", "");
}

export function closure3(unitVar: void, v0_1: string): Date {
    return parse_1(closure4(undefined, v0_1.slice(undefined, 24 + 1)));
}

export function closure7(v0_1: int32, v1_1: string, v2_1: int32, v3_1: string): string {
    if (v2_1 >= v0_1) {
        return v3_1;
    }
    else {
        return method5(v0_1, v1_1, v2_1 + 1)(v3_1 + v1_1);
    }
}

export function method5(v0_1: int32, v1_1: string, v2_1: int32): ((arg0: string) => string) {
    return (v: string): string => closure7(v0_1, v1_1, v2_1, v);
}

export function closure6(v0_1: string, v1_1: int64): string {
    const v3_1: string = v0_1;
    const v7_1: string = int64ToString(v1_1);
    const v11 = "0";
    const v20: string = method5(18 - v7_1.length, v11, 0)("") + v7_1;
    return parse(`${v20.slice(0, 7 + 1)}-${v20.slice(8, 11 + 1)}-${v20.slice(12, 15 + 1)}-${v20.slice(16, 17 + 1)}${v3_1.slice(21, v3_1.length)}`);
}

export function closure5(unitVar: void, v0_1: string): ((arg0: int64) => string) {
    return (v: int64): string => closure6(v0_1, v);
}

export function closure8(unitVar: void, v0_1: string): int64 {
    const v2_1: string = v0_1;
    return toInt64(parse_2(concat(v2_1.slice(0, 7 + 1), v2_1.slice(9, 12 + 1), v2_1.slice(14, 17 + 1), ...v2_1.slice(19, 20 + 1)), 511, false, 64));
}

export function closure9(unitVar: void, v0_1: Date): string {
    return method0(newGuid(), v0_1);
}

export function closure10(unitVar: void, v0_1: int64): string {
    const v6_1: string = newGuid();
    const v10: string = int64ToString(v0_1);
    const v14 = "0";
    const v23: string = method5(18 - v10.length, v14, 0)("") + v10;
    return parse(`${v23.slice(0, 7 + 1)}-${v23.slice(8, 11 + 1)}-${v23.slice(12, 15 + 1)}-${v23.slice(16, 17 + 1)}${v6_1.slice(21, v6_1.length)}`);
}

export function closure12(v0_1: string, v1_1: Date): string {
    return toString(v1_1, (v0_1 === "") ? "M-d-y hh:mm:ss tt" : v0_1);
}

export function closure11(unitVar: void, v0_1: string): ((arg0: Date) => string) {
    return (v: Date): string => closure12(v0_1, v);
}

export function closure13(unitVar: void, v0_1: Date): string {
    return toString(v0_1, "yyyy-MM-ddTHH-mm-ss.fff");
}

export const v0 = (v: string): ((arg0: Date) => string) => closure0(undefined, v);

export function date_time_guid_from_date_time(x: string): ((arg0: Date) => string) {
    return v0(x);
}

export const v1 = (v: string): Date => closure3(undefined, v);

export function date_time_from_guid(x: string): Date {
    return v1(x);
}

export const v2 = (v: string): ((arg0: int64) => string) => closure5(undefined, v);

export function timestamp_guid_from_timestamp(x: string): ((arg0: int64) => string) {
    return v2(x);
}

export const v3 = (v: string): int64 => closure8(undefined, v);

export function timestamp_from_guid(x: string): int64 {
    return v3(x);
}

export const v4 = (v: Date): string => closure9(undefined, v);

export function new_guid_from_date_time(x: Date): string {
    return v4(x);
}

export const v5 = (v: int64): string => closure10(undefined, v);

export function new_guid_from_timestamp(x: int64): string {
    return v5(x);
}

export const v6 = (v: string): ((arg0: Date) => string) => closure11(undefined, v);

export function format(x: string): ((arg0: Date) => string) {
    return v6(x);
}

export const v7 = (v: Date): string => closure13(undefined, v);

export function format_iso8601(x: Date): string {
    return v7(x);
}

File: lib/spiral/date_time.py
from __future__ import annotations
from collections.abc import Callable
from typing import Any
from fable_modules.fable_library.date import (to_string, parse_exact)
from fable_modules.fable_library.guid import (parse, new_guid)
from fable_modules.fable_library.long import parse as parse_1
from fable_modules.fable_library.reflection import (TypeInfo, class_type, union_type, string_type)
from fable_modules.fable_library.string_ import (concat, replace)
from fable_modules.fable_library.time_span import (hours, to_string as to_string_1)
from fable_modules.fable_library.types import (Array, Union, uint8, int64)
from fable_modules.fable_library.util import int64_to_string

def _expr108() -> TypeInfo:
    return union_type("Date_time.US0", [], US0, lambda: [[("f0_0", class_type("Date_time.chrono_DateTime`1", [class_type("Date_time.chrono_Utc")]))], []])


class US0(Union):
    __slots__ = ["tag", "fields"]
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["US0_0", "US0_1"]


US0_reflection = _expr108

def _expr109() -> TypeInfo:
    return union_type("Date_time.US1", [], US1, lambda: [[("f0_0", string_type)], []])


class US1(Union):
    __slots__ = ["tag", "fields"]
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["US1_0", "US1_1"]


US1_reflection = _expr109

def _expr110() -> TypeInfo:
    return union_type("Date_time.US2", [], US2, lambda: [[], [], []])


class US2(Union):
    __slots__ = ["tag", "fields"]
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["US2_0", "US2_1", "US2_2"]


US2_reflection = _expr110

def _expr111() -> TypeInfo:
    return union_type("Date_time.US3", [], US3, lambda: [[("f0_0", US2_reflection())], [("f1_0", US2_reflection())], [("f2_0", US2_reflection())], [("f3_0", US2_reflection())], [("f4_0", US2_reflection())]])


class US3(Union):
    __slots__ = ["tag", "fields"]
    def __init__(self, tag: int, *fields: Any) -> None:
        super().__init__()
        self.tag: int = tag or 0
        self.fields: Array[Any] = list(fields)

    @staticmethod
    def cases() -> list[str]:
        return ["US3_0", "US3_1", "US3_2", "US3_3", "US3_4"]


US3_reflection = _expr111

def closure2(unit_var: None, v0_1: chrono_date_time_1[chrono_utc]) -> US0:
    return US0(0, v0_1)


def method1(__unit: None=None) -> Callable[[chrono_date_time_1[chrono_utc]], US0]:
    def _arrow112(v: chrono_date_time_1[chrono_utc]) -> US0:
        return closure2(None, v)

    return _arrow112


def method2(__unit: None=None) -> str:
    return "hh:mm"


def method3(__unit: None=None) -> str:
    return "yyyyMMdd-HHmm-ssff-ffff-f"


def method4(__unit: None=None) -> str:
    return "hhmm"


def method0(v0_1: str, v1_1: Any) -> str:
    v1075: str = method3()
    v1080: str = to_string(v1_1, "M-d-y hh:mm:ss tt" if (v1075 == "") else v1075)
    v1261: Any
    raise Exception(("date_time.get_utc_offset / target: " + str(US3(4, US2(0)))) + "")
    v1274: uint8 = uint8(1) if (hours(v1261) > 0) else uint8(0)
    v1275: str = method4()
    v1299: str = ((("" + str(v1274)) + "") + to_string_1(v1261, "c", {})) + ""
    v1301: str = str(v0_1)
    v1306: int = (len(v1080) + len(v1299)) or 0
    v1325: int = (len(v1301) - 1) or 0
    return parse(concat(v1080, v1299, *v1301[v1306:v1325 + 1]))


def closure1(v0_1: str, v1_1: Any) -> str:
    return method0(v0_1, v1_1)


def closure0(unit_var: None, v0_1: str) -> Callable[[Any], str]:
    def _arrow113(v: Any, unit_var: Any=unit_var, v0_1: Any=v0_1) -> str:
        return closure1(v0_1, v)

    return _arrow113


def closure4(unit_var: None, v0_1: str) -> str:
    return replace(v0_1, "-", "")


def closure3(unit_var: None, v0_1: str) -> Any:
    v2_1: str = str(v0_1)
    return parse_exact(closure4(None, v2_1[:24 + 1]), "yyyyMMddHHmmssfffffff", None)


def closure7(v0_1: int, v1_1: str, v2_1: int, v3_1: str) -> str:
    if v2_1 >= v0_1:
        return v3_1

    else: 
        return method5(v0_1, v1_1, v2_1 + 1)(v3_1 + v1_1)



def method5(v0_1: int, v1_1: str, v2_1: int) -> Callable[[str], str]:
    def _arrow114(v: str, v0_1: Any=v0_1, v1_1: Any=v1_1, v2_1: Any=v2_1) -> str:
        return closure7(v0_1, v1_1, v2_1, v)

    return _arrow114


def closure6(v0_1: str, v1_1: int64) -> str:
    v3_1: str = str(v0_1)
    v7_1: str = int64_to_string(v1_1)
    v20: str = method5(18 - len(v7_1), "0", 0)("") + v7_1
    return parse(((((((((("" + v20[0:7 + 1]) + "-") + v20[8:11 + 1]) + "-") + v20[12:15 + 1]) + "-") + v20[16:17 + 1]) + "") + v3_1[21:len(v3_1)]) + "")


def closure5(unit_var: None, v0_1: str) -> Callable[[int64], str]:
    def _arrow115(v: int64, unit_var: Any=unit_var, v0_1: Any=v0_1) -> str:
        return closure6(v0_1, v)

    return _arrow115


def closure8(unit_var: None, v0_1: str) -> int64:
    v2_1: str = str(v0_1)
    return parse_1(concat(v2_1[0:7 + 1], v2_1[9:12 + 1], v2_1[14:17 + 1], *v2_1[19:20 + 1]), 511, False, 64)


def closure9(unit_var: None, v0_1: Any) -> str:
    return method0(new_guid(), v0_1)


def closure10(unit_var: None, v0_1: int64) -> str:
    v6_1: str = str(new_guid())
    v10: str = int64_to_string(v0_1)
    v23: str = method5(18 - len(v10), "0", 0)("") + v10
    return parse(((((((((("" + v23[0:7 + 1]) + "-") + v23[8:11 + 1]) + "-") + v23[12:15 + 1]) + "-") + v23[16:17 + 1]) + "") + v6_1[21:len(v6_1)]) + "")


def closure12(v0_1: str, v1_1: Any) -> str:
    return to_string(v1_1, "M-d-y hh:mm:ss tt" if (v0_1 == "") else v0_1)


def closure11(unit_var: None, v0_1: str) -> Callable[[Any], str]:
    def _arrow116(v: Any, unit_var: Any=unit_var, v0_1: Any=v0_1) -> str:
        return closure12(v0_1, v)

    return _arrow116


def closure13(unit_var: None, v0_1: Any) -> str:
    return to_string(v0_1, "yyyy-MM-ddTHH-mm-ss.fff")


def _arrow117(v: str) -> Callable[[Any], str]:
    return closure0(None, v)


v0: Callable[[str, Any], str] = _arrow117

def date_time_guid_from_date_time(x: str) -> Callable[[Any], str]:
    return v0(x)


def _arrow118(v: str) -> Any:
    return closure3(None, v)


v1: Callable[[str], Any] = _arrow118

def date_time_from_guid(x: str) -> Any:
    return v1(x)


def _arrow119(v: str) -> Callable[[int64], str]:
    return closure5(None, v)


v2: Callable[[str, int64], str] = _arrow119

def timestamp_guid_from_timestamp(x: str) -> Callable[[int64], str]:
    return v2(x)


def _arrow120(v: str) -> int64:
    return closure8(None, v)


v3: Callable[[str], int64] = _arrow120

def timestamp_from_guid(x: str) -> int64:
    return v3(x)


def _arrow121(v: Any) -> str:
    return closure9(None, v)


v4: Callable[[Any], str] = _arrow121

def new_guid_from_date_time(x: Any) -> str:
    return v4(x)


def _arrow122(v: int64) -> str:
    return closure10(None, v)


v5: Callable[[int64], str] = _arrow122

def new_guid_from_timestamp(x: int64) -> str:
    return v5(x)


def _arrow123(v: str) -> Callable[[Any], str]:
    return closure11(None, v)


v6: Callable[[str, Any], str] = _arrow123

def format(x: str) -> Callable[[Any], str]:
    return v6(x)


def _arrow124(v: Any) -> str:
    return closure13(None, v)


v7: Callable[[Any], str] = _arrow124

def format_iso8601(x: Any) -> str:
    return v7(x)


